### PR TITLE
Alchemy Plugin

### PR DESCRIPTION
**Pull request by samholmes**
### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [x] Yes
- [ ] No

### Dependencies

- https://github.com/EdgeApp/edge-change-server/pull/12

### Description

 none

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1213206026575421
  - https://app.asana.com/0/0/1212754123530597

---

> [!NOTE]
> **High Risk**
> Adds a new externally reachable webhook HTTP surface and signature validation/auth logic, plus cluster-wide IPC relaying; failures here could drop or misroute activity updates or weaken webhook authentication.
> 
> **Overview**
> Adds an Alchemy-based address activity pipeline: workers now host an Express HTTP endpoint for `/webhook/*` (with GET/HEAD health checks) and dispatch incoming webhooks through a new `WebhookRegistry` to per-plugin handlers.
> 
> Introduces a new `alchemy` plugin that creates/updates/deletes Alchemy Address Activity webhooks, validates `X-Alchemy-Signature` via HMAC using recovered signing keys, debounces subscription address updates, and broadcasts webhook activity across cluster workers via IPC so all workers emit updates consistently.
> 
> Updates configuration/deps to support this (new `webhookHost`/`webhookPort` and `alchemyAuthToken`, adds `express` + types, bumps `cleaner-config`), and wires plugin construction through `makeAllPlugins(opts)` with Alchemy enabled for multiple EVM networks plus Solana; includes comprehensive Jest tests for the plugin behavior and signature/auth paths.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b3d41ce085fff00b5e0c5f8d00bbf52c79716879. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/util/webhookRegistry.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+          const response = await routeHandler(request)
+          const status = response.status ?? 200
+          if (status < 200 || status >= 300) {
+            return response
+          }
+        } catch (err: unknown) {
+          logger.error({ err, webhookKey, msg: 'Error in webhook route' })
+        }
+      }
+      return OK_RESPONSE
```
Comment (paullinator) [inline src/util/webhookRegistry.ts:77]: **Critical: Swallowed exceptions return 200 OK to Alchemy, preventing retries**

When a handler throws (e.g., `signingKeyStore.getSigningKey()` throws during key recovery), the catch logs the error but falls through to return `200 OK`. Alchemy considers the webhook delivered and will not retry. Combined with the signing key store re-throwing on transient API failure, a temporary outage causes webhooks to silently succeed without processing.

**Recommendation:** Return a 500 response on unexpected exceptions so Alchemy retries:

```typescript
} catch (err: unknown) {
  logger.error({ err, webhookKey, msg: 'Error in webhook route' })
  return {
    status: 500,
    headers: { 'content-type': 'text/plain' },
    body: 'Internal error'
  }
}
```

File: src/plugins/alchemy.ts [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+          : undefined
+      emit('update', { address, checkpoint })
+    }
+  }
+
+  /**
+   * Process batched address changes
+   */
+  async function processBatchedChanges(): Promise<void> {
+    batchTimeout = null
```
Comment (paullinator) [inline src/plugins/alchemy.ts:236]: **Critical: Race condition in concurrent batch processing**

`processBatchedChanges` is async and sets `batchTimeout = null` at the top. While it's awaiting API calls (lines 233-266), `scheduleBatch()` can be called again — from the error retry on line 287 or from a new `subscribe()` call — scheduling a second concurrent execution. Two concurrent runs could both see `webhookId == null` and each try to `createWebhook`, creating duplicate webhooks.

**Recommendation:** Add a processing guard:

```typescript
let processing = false

async function processBatchedChanges(): Promise<void> {
  if (processing) {
    scheduleBatch()
    return
  }
  processing = true
  try {
    batchTimeout = null
    // ...existing logic...
  } finally {
    processing = false
  }
}
```

File: src/plugins/alchemy.ts [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+    } catch (err: unknown) {
+      logger.error({ err, msg: 'Failed to update webhook addresses' })
+
+      // Re-queue failed operations
+      pendingAddressesToAdd.push(...toAdd)
+      pendingAddressesToRemove.push(...toRemove)
+
+      // Retry after delay
+      scheduleBatch()
+    }
```
Comment (paullinator) [inline src/plugins/alchemy.ts:314]: **Critical: No retry backoff — tight 1-second retry loop when Alchemy API is down**

When `processBatchedChanges` fails, it re-queues addresses and calls `scheduleBatch()`, which schedules a new attempt in `BATCH_DELAY_MS` (1 second). If the Alchemy API is unreachable, this creates an infinite 1-second polling loop hammering the API.

**Recommendation:** Implement exponential backoff:

```typescript
let retryCount = 0
const MAX_RETRY_DELAY_MS = 60_000

// In the catch:
retryCount++
const delay = Math.min(BATCH_DELAY_MS * Math.pow(2, retryCount), MAX_RETRY_DELAY_MS)
scheduleBatch(delay)

// Reset retryCount = 0 on success
```

File: src/plugins/alchemy.ts [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+      webhookRegistry.unregisterHandler(WEBHOOK_KEY, webhookRoute)
+
+      // Clear local state
+      subscribedAddresses.clear()
+      pendingAddressesToAdd = []
+      pendingAddressesToRemove = []
+
+      // Note: We don't delete the webhook on destroy
+      // It can be reused on restart, and addresses can be managed
+    }
```
Comment (paullinator) [inline src/plugins/alchemy.ts:444]: **Critical: `destroy()` doesn't cancel in-flight `processBatchedChanges()`**

When the batch timeout fires, `processBatchedChanges()` runs asynchronously. If `destroy()` is called while it's mid-flight (after the timeout fired but before awaits resolve), it clears `subscribedAddresses` and pending arrays. But the in-flight function holds local references (`toAdd`, `toRemove`). On error (line 283), it pushes addresses back onto the now-cleared arrays and calls `scheduleBatch()`, re-creating a timeout on a destroyed plugin.

**Recommendation:** Add a `destroyed` flag:

```typescript
let destroyed = false

async function processBatchedChanges(): Promise<void> {
  batchTimeout = null
  if (destroyed) return
  // ...existing code...
  } catch (err: unknown) {
    if (destroyed) return
    // Re-queue + retry...
  }
}

destroy() {
  destroyed = true
  // ...existing cleanup...
}
```

File: src/plugins/alchemy.ts [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+      initialized = true
+      logger.info({
+        webhookId,
+        msg: 'Webhook discovery complete'
+      })
+    } catch (err: unknown) {
+      logger.error({ err, msg: 'Failed to discover webhooks' })
+      // Mark as initialized anyway to prevent blocking
+      initialized = true
+    }
```
Comment (paullinator) [inline src/plugins/alchemy.ts:192]: **Warning: Initialization failure permanently prevents webhook discovery**

If `doInitialize` fails (network error, API outage), `initialized` is set to `true` permanently. The plugin will never attempt discovery again on subsequent calls to `initialize()`, meaning it will try to create new webhooks instead of reusing existing ones.

**Recommendation:** Consider leaving `initialized = false` with a cooldown timer, or retrying on subsequent `processBatchedChanges()` calls.

File: src/util/signingKeyStore.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+      // If not in cache and we haven't tried recovery, do it now
+      if (!hasRecovered) {
+        logger.info({
+          webhookId,
+          msg: 'Signing key not cached, recovering from API'
+        })
+        await this.recoverSigningKeys()
+        signingKey = signingKeys.get(webhookId)
+      }
```
Comment (paullinator) [inline src/util/signingKeyStore.ts]: **Warning: Concurrent signing key recovery race condition**

If two webhook requests arrive simultaneously for unknown webhook IDs, both check `hasRecovered` (false on line 50) and both call `this.recoverSigningKeys()`, triggering duplicate `getTeamWebhooks()` API calls. Unlike `alchemy.ts` which correctly uses an `initializingPromise` singleton pattern, the signing key store has no deduplication.

Additionally, `recoverSigningKeys()` re-throws on failure (line 82). When called from `getSigningKey()`, this propagates the exception to the webhook route handler, which (combined with the webhookRegistry issue) causes the webhook to silently succeed with 200 OK without processing.

**Recommendation:** Use the same promise-singleton pattern:

```typescript
let recoveringPromise: Promise<void> | null = null

async getSigningKey(webhookId: string): Promise<string | undefined> {
  // ...cache check...
  if (!hasRecovered) {
    if (recoveringPromise == null) {
      recoveringPromise = this.recoverSigningKeys()
        .finally(() => { recoveringPromise = null })
    }
    try {
      await recoveringPromise
    } catch {
      // Return undefined; webhook handler already handles null key with 401
    }
    signingKey = signingKeys.get(webhookId)
  }
  // ...
}
```

File: src/plugins/alchemy.ts [hunk 7]
```diff
@@ -0,0 +1,579 @@
...
+    handleActivity
+  )
+
+  // Register this plugin's handler with the webhook registry
+  webhookRegistry.registerHandler(WEBHOOK_KEY, webhookRoute)
+
+  // Initialize immediately on plugin creation (cleanup paused webhooks, discover existing)
+  initialize().catch((err: unknown) => {
+    logger.error({ err, msg: 'Failed to initialize on startup' })
+  })
```
Comment (paullinator) [inline src/plugins/alchemy.ts:378]: **Warning: 14 concurrent `getTeamWebhooks()` calls on startup**

Each of the 14 `makeAlchemy()` calls in `allPlugins.ts` fires `initialize()` immediately, and each calls `notifyApi.getTeamWebhooks()` concurrently. This results in 14 simultaneous API calls to the same Alchemy endpoint, which is wasteful and risks rate-limiting.

**Recommendation:** Share the `getTeamWebhooks()` result across all plugin instances, e.g. via a shared initialization promise or by passing the pre-fetched webhook list into `makeAlchemy()`.

File: src/util/alchemyNotifyApi.ts [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+    const text = await response.text()
+    if (asResponse != null) {
+      try {
+        return asResponse(text)
+      } catch (err: unknown) {
+        logger.error({ err, responseText: text }, 'Alchemy API cleaner error')
+        throw err
+      }
+    }
+    return (undefined as unknown) as T
```
Comment (paullinator) [inline src/util/alchemyNotifyApi.ts]: **Warning: Unsafe `(undefined as unknown) as T` type casts defeat type safety**

Two paths return `undefined` cast to `T`. If a caller expects typed data from these paths, it will get `undefined` at runtime with no compile-time warning. Also, line 253 uses the Pino positional `(obj, 'msg')` form while all other log calls in the PR use the `{ msg: '' }` object pattern — inconsistent.

**Recommendation:** Split into separate functions or use overloads:

```typescript
async function apiRequestVoid(endpoint: string, method: string, body?: object): Promise<void> { ... }
async function apiRequest<T>(endpoint: string, method: string, body: object | undefined, asResponse: Cleaner<T>): Promise<T> { ... }
```

For the logger, use the consistent `{ msg: 'Alchemy API cleaner error' }` pattern.

File: src/plugins/alchemy.ts [hunk 9]
```diff
@@ -0,0 +1,579 @@
...
+      subscribedAddresses.set(normalizedAddress, address)
+
+      // Queue for batch update to Alchemy
+      pendingAddressesToAdd.push(address)
+
+      // Remove from pending removals if present
+      const removeIndex = pendingAddressesToRemove.indexOf(address)
+      if (removeIndex !== -1) {
+        pendingAddressesToRemove.splice(removeIndex, 1)
+      }
```
Comment (paullinator) [inline src/plugins/alchemy.ts]: **Suggestion: Pending arrays use original casing; cross-cased subscribe/unsubscribe could leave phantom entries**

`subscribe` pushes `address` (original casing) to `pendingAddressesToAdd`. `unsubscribe` then uses `pendingAddressesToAdd.indexOf(address)` on line 366 to cancel it. If the same address is passed with different casing between `subscribe` and `unsubscribe`, `indexOf` won't match, leaving a phantom entry in the pending add list.

**Recommendation:** Use the normalized (lowercase) address consistently in the pending arrays, or search using a case-insensitive comparison.

File: src/plugins/alchemy.ts [hunk 10]
```diff
@@ -0,0 +1,579 @@
...
+    }
+
+    // Emit update events for all affected subscribed addresses
+    for (const address of addressesToUpdate) {
+      // Use the block number from the first activity as checkpoint
+      const checkpoint =
+        activities.length > 0
+          ? parseInt(activities[0].blockNum, 16).toString()
+          : undefined
```
Comment (paullinator) [inline src/plugins/alchemy.ts]: **Suggestion: `parseInt` on hex block number can produce NaN checkpoint**

If `blockNum` is a string that doesn't parse as hex (the cleaner only validates it's a string, not a valid hex number), `parseInt` returns `NaN`, and `toString()` produces the string `"NaN"` as a checkpoint.

Also, when a payload contains multiple activities spanning different blocks, the checkpoint is always from `activities[0].blockNum`. If a later activity is in a higher block, the checkpoint underreports progress.

**Recommendation:**
```typescript
const blockNumber = parseInt(activities[0].blockNum, 16)
const checkpoint = Number.isNaN(blockNumber) ? undefined : blockNumber.toString()
```

File: test/plugins/alchemy.test.ts [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+      headers: { 'x-alchemy-signature': signature },
+      req: { body } as any
+    })
+  }
+
+  beforeEach(() => {
+    jest.clearAllMocks()
+    jest.useFakeTimers()
+    registeredHandler = null
+
```
Comment (paullinator) [inline test/plugins/alchemy.test.ts:128]: **Critical: Batch processing (the core Alchemy integration) is never exercised in tests**

The tests use `jest.useFakeTimers()` here but never advance them with `jest.advanceTimersByTimeAsync(1000)`. Since `subscribe()` only queues changes and calls `scheduleBatch()` (which sets a `setTimeout`), the actual `processBatchedChanges()` function never fires. This means:

- `notifyApi.createWebhook` is never called
- `notifyApi.updateWebhookAddresses` is never called
- `notifyApi.deleteWebhook` is never called (for empty subscriptions)
- Webhook lifecycle (creation, address syncing, deletion) is **completely untested**

Additionally, signature validation rejection paths (invalid/missing signature, unknown webhook ID) have zero test coverage — this is the security boundary.

**Recommendation:** Add tests like:
```typescript
test('subscribe should create webhook after batch delay', async () => {
  await plugin.subscribe(TEST_ADDRESS)
  await jest.advanceTimersByTimeAsync(1000)
  expect(notifyApi.createWebhook).toHaveBeenCalledWith({
    network: 'ETH_MAINNET',
    webhookUrl: 'https://test.edge.app/webhook/alchemy/ethereum',
    addresses: [TEST_ADDRESS]
  })
})

test('invalid signature should return 401', async () => {
  // Call handler with wrong signature...
})
```

**Review by paullinator — CHANGES_REQUESTED** [review]
## Review Summary

The overall architecture is well-designed — clean separation between plugin, API client, signing key store, and webhook registry. The timing-safe HMAC signature validation is solid, and the debounced batch address management is a good approach.

However, there are several **concurrency and error handling issues** that could cause silent data loss in production:

### Critical Issues (4)
1. **Swallowed exceptions in webhookRegistry** return 200 OK to Alchemy, preventing retries
2. **Race condition** in concurrent `processBatchedChanges()` — can create duplicate webhooks
3. **No retry backoff** — tight 1-second loop hammering Alchemy API during outages
4. **`destroy()` doesn't cancel in-flight batch processing** — can re-create timers on destroyed plugin

### Critical Test Gap (1)
5. **Batch processing is never exercised in tests** — fake timers are never advanced, so the most important code path (Alchemy API integration) has zero test coverage. Signature validation rejection paths are also untested.

### Warnings (4)
6. Initialization failure permanently prevents webhook discovery
7. Concurrent signing key recovery race condition (no promise deduplication)
8. 14 concurrent `getTeamWebhooks()` API calls on startup
9. Unsafe `(undefined as unknown) as T` type casts in API client

### Suggestions (2)
10. Cross-cased subscribe/unsubscribe can leave phantom entries in pending arrays
11. `parseInt` on hex block number can produce NaN checkpoint

**Note:** Cursor Bugbot already flagged the cluster multi-worker affinity issue, network-check-before-auth ordering, and @types/express placement — those are omitted from this review.

See inline comments for specific details and recommended fixes.

File: src/plugins/allPlugins.ts [hunk 12]
```diff
@@ -1,33 +1,45 @@
-import { makeBlockbook } from './blockbook'
+import { notifyApi, signingKeyStore, webhookRegistry } from '../services'
+import { AddressPlugin } from '../types/addressPlugin'
+import { makeAlchemy } from './alchemy'
 import { makeEvmRpc } from './evmRpc'
 import { makeFakePlugin } from './fakePlugin'
 
-export const allPlugins = [
+const teamWebhooksPromise = notifyApi.getTeamWebhooks()
```
[the remark this answers is not in the document]
Comment (swansontec) [reply src/plugins/allPlugins.ts]: I agree. Would something like this work?

```ts
const teamWebhooksPromise: Promise<...> | undefined = undefined

makeAlchemyPlugin() {
  // ...
  if (teamWebhooksPromise == null) {
    teamWebhooksPromise = notifyApi.getTeamWebhooks()
  }
  await teamWebhooksPromise.catch(() => { teamWebhooksPromise = undefined })
```

File: src/plugins/allPlugins.ts [hunk 13]
```diff
@@ -34,59 +39,119 @@ export const allPlugins = [
...
+    pluginId: 'rsk',
+    network: 'ROOTSTOCK_MAINNET',
+    notifyApi,
+    signingKeyStore,
+    webhookRegistry,
+    teamWebhooksPromise
+  }),
+  makeAlchemy({
+    pluginId: 'solana',
+    network: 'SOLANA_MAINNET',
```
Comment (swansontec) [inline src/plugins/allPlugins.ts]: Is this tested compatible with Solana? We do `const normalizedAddress = address.toLowerCase()`, which doesn't work with base58.

File: src/plugins/alchemy.ts [hunk 14]
```diff
@@ -0,0 +1,618 @@
...
+  } = opts
+
+  const WEBHOOK_KEY = `alchemy/${pluginId}`
+
+  const [on, emit] = makeEvents<PluginEvents>()
+
+  const logger: Logger = makeLogger('alchemy', pluginId)
+
+  // Track subscribed addresses (normalized lowercase address -> original address)
+  const subscribedAddresses = new Map<string, string>()
```
Comment (swansontec) [inline src/plugins/alchemy.ts:64]: As mentioned, this normalization doesn't really work for base58 addresses. Can we come up with a design that doesn't require normalization?

File: src/services.ts [hunk 15]
```diff
@@ -0,0 +1,8 @@
+import { makeAlchemyNotifyApi } from './util/alchemyNotifyApi'
+import { makeSigningKeyStore } from './util/signingKeyStore'
+import { makeWebhookRegistry } from './util/webhookRegistry'
+
+// Create singleton instances
+export const notifyApi = makeAlchemyNotifyApi()
+export const signingKeyStore = makeSigningKeyStore({ notifyApi })
+export const webhookRegistry = makeWebhookRegistry()
```
Comment (swansontec) [inline src/services.ts]: I am not really a fan of this design, especially now that we have multiple process and IPC involved. I would rather have these things be dependency-injected into whoever needs them. This may involve some re-shuffling, but mostly in setup code.

For instance, `export const allPlugins: AddressPlugin[] = [...]` might need to turn into an `export const getPlugins: ({ services... }) => ([...])`. Then the various top-level processes can be in charge of setting up singletons and passing them in where they belong. This would also get rid of the need for `await import('./server')`, as the server would no longer have global state floating around.

File: src/server.ts
Comment (swansontec) [inline src/server.ts:1]: I think "server.ts" is a weird name. Which server is this? The websocket server for the core, or the webhook server for Alchemy? I'd like to have it named "websocketServer.ts"

**Review by swansontec — CHANGES_REQUESTED** [review]

File: src/server.ts
Comment (samholmes) [reply src/server.ts:1]: worker.ts would be better. The file contains a websocket server and http server for the webhook.

File: src/services.ts (hunk 15 above)
Comment (samholmes) [reply src/services.ts]: Each process gets it's own singleton. The coupling of these services to the rest of the app is simpler to reason about (we're only having one instance of each service as part of the worker process anyway). I can see how dep-injection could make sense here though.

If we want to remove `await import('./server')` then we'd also have to call a `server()` function that is exported form `./server.ts`.

File: src/plugins/allPlugins.ts (hunk 13 above)
Comment (samholmes) [reply src/plugins/allPlugins.ts]: Client-side we have no chagne-server integration with Solana. No testing done with Solana, but I'll make sure to make normalization an option for the plugins rather then default.

File: src/index.ts [hunk 16]
```diff
@@ -68,7 +77,4 @@ function manageServers(): void {
   logger.info({ port: metricsPort }, 'metrics server listening')
 }
 
-main().catch(error => {
-  logger.error({ err: error }, 'main error')
-  process.exit(1)
-})
```
Comment (swansontec) [inline src/index.ts:104]: I would leave this catch statement in place, since "the buck starts here". This is our last chance to log if anything goes wrong.

**Review by swansontec — COMMENTED** [review]
One change, but I'm approving because it's trivial.

File: src/index.ts (hunk 16 above)
Comment (samholmes) [reply src/index.ts:104]: The function changed from async to sync
